From 70254532223d303c67cf077270edfbdf98cb03de Mon Sep 17 00:00:00 2001
From: Jose Dapena Paz <jose.dapena@lge.com>
Date: Thu, 1 Mar 2018 17:56:12 +0100
Subject: GCC build fix: mark is_trivially_copy_constructible for WTF::Vector
 as false.

Compilation in GCC fails because of https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80654
if T in WTF::Optional<WTF::Vector<T>> is not trivially copy constructible.
The problem already happened in std::vector and was workarounded. This
change implements a similar fix for WTF::Vector.

It creates a WTF TemplateUtil.h just in case we need to add more template workarounds
in the future.

Bug: 816952

Change-Id: If87f01beb952e03eb49dcaf0c5db6efd745bf05e

diff --git a/third_party/WebKit/Source/platform/wtf/Optional.h b/third_party/WebKit/Source/platform/wtf/Optional.h
index c5a6b67fdb3c..4e28848ff8a7 100644
--- a/third_party/WebKit/Source/platform/wtf/Optional.h
+++ b/third_party/WebKit/Source/platform/wtf/Optional.h
@@ -6,6 +6,7 @@
 #define Optional_h
 
 #include "base/optional.h"
+#include "platform/wtf/TemplateUtil.h"
 
 namespace WTF {
 
diff --git a/third_party/WebKit/Source/platform/wtf/TemplateUtil.h b/third_party/WebKit/Source/platform/wtf/TemplateUtil.h
new file mode 100644
index 000000000000..ad27afc5f7f7
--- /dev/null
+++ b/third_party/WebKit/Source/platform/wtf/TemplateUtil.h
@@ -0,0 +1,27 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef TemplateUtil_h
+#define TemplateUtil_h
+
+#include "base/template_util.h"
+#include "platform/wtf/Vector.h"
+
+namespace base {
+
+#if defined(__GNUC__) && !defined(__clang__) && __GNUC__ <= 7
+// Workaround for g++7 and earlier family.
+// Due to https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80654, without this
+// Optional<WTF::Vector<T>> where T is non-copyable causes a compile error.
+// As we know it is not trivially copy constructible, explicitly declare so.
+//
+// It completes the declaration in base/template_util.h that was provided
+// for std::vector
+template <typename T>
+struct is_trivially_copy_constructible<WTF::Vector<T>> : std::false_type {};
+#endif
+
+}  // namespace base
+
+#endif  // TemplateUtil_h
-- 
2.14.1

